cloudrun-malware-scanner/zip.js

```diff
@@ -1,35 +1,219 @@
 const {logger} = require('./logger.js');
 const unzip = require('unzip-stream');
+const clamd = require('clamdjs');
+
+const CLAMD_MAX_FILE_SIZE = 4 * 1024 * 1024 * 1024; // 4GB
+const SIZE_LIMIT_ERROR = 'INSTREAM size limit exceeded';
+const FALLBACK_RESOLVE_TIMEOUT = 10 * 1000; // 10 seconds
+const CHUNK_SCAN_OVERLAP_FACTOR = 0.1;
 
 /**
  * @param {*} scanner
- * @param {*} readstream
- * @param {number} totalSize
+ * @param {*} gcsFile
  * @param {number} timeoutPerFile
- * @return {Promise<Array<string>>} collection of clamd results
+ * @param {number} chunkSize
+ * @return {Promise<Array<*>>} collection of clamd results
  */
-function scanGcsZipFile(scanner, readstream, totalSize, timeoutPerFile) {
-  const results = [];
+function scanGcsZipFile(scanner, gcsFile, timeoutPerFile, chunkSize) {
+  const scanPromises = [];
+
+  let fileCount = 0;
+  let scannedCount = 0;
 
   return new Promise((resolve, reject) => {
+    const readstream = gcsFile.createReadStream();
+
+    // First, try to scan zip archive fully and scan each individual file in
+    // full with clamd. If that fails, then break larger files up into smaller
+    // chunks to scan to avoid clam's hard 4GB limit.
     // eslint-disable-next-line new-cap
     readstream.pipe(unzip.Parse()).on('entry', async (entry) => {
       if (entry.type === 'File') {
-        logger.info(`Scanning ${entry.path} with size ${entry.size}`);
-        const result = await scanner.scanStream(entry, timeoutPerFile);
-        results.push({result, fileName: entry.path});
-        logger.info(`Finished scanning ${entry.path} with result ${result}`);
+        logger.info(`
```

Comment:
ideally, we have a stream transformation that just does:

zipFile -> ... bunch of transformers ... -> stream(chunks)
which we feed to scanner